packages/java/endpoint/src/main/java/dev/hilla/crud/CrudRepositoryService.java

```diff
@@ -1,49 +1,104 @@
 package dev.hilla.crud;
 
+import java.lang.reflect.Type;
 import java.util.List;
 
+import com.googlecode.gentyref.GenericTypeReflector;
+
 import dev.hilla.EndpointExposed;
 import dev.hilla.Nullable;
 import dev.hilla.crud.filter.Filter;
+
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationContext;
 import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.domain.Specification;
-import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.repository.CrudRepository;
 
 /**
  * A browser-callable service that delegates crud operations to a JPA
  * repository.
  */
 @EndpointExposed
-public class CrudRepositoryService<T, ID> implements CrudService<T> {
+public class CrudRepositoryService<T, ID, R extends CrudRepository<T, ID> & JpaSpecificationExecutor<T>>
+        implements CrudService<T> {
 
     @Autowired
     private JpaFilterConverter jpaFilterConverter;
 
-    private JpaSpecificationExecutor<T> repository;
-    private Class<T> entityClass;
+    @Autowired
+    private ApplicationContext applicationContext;
+
+    private R repository;
+    private final Class<T> entityClass;
+
+    public CrudRepositoryService() {
+        this.entityClass = resolveEntityClass();
+    }
 
     /**
      * Creates the service using the given repository.
      *
      * @param repository
      *            the JPA repository
      */
-    public <R extends JpaRepository<T, ID> & JpaSpecificationExecutor<T>> CrudRepositoryService(
-            Class<T> entityClass, R repository) {
+    public CrudRepositoryService(R repository) {
         this.repository = repository;
-        this.entityClass = entityClass;
+        this.entityClass = resolveEntityClass();
     }
 
-    protected JpaSpecificationExecutor<T> getRepository() {
+    /**
+     * Initializes the repository instance if it hasn't been initialized yet,
+     * otherwise returns the existing one.
+     * <p>
+     * This method uses ApplicationContext to obtain the Repository instance, so
+     * it is not suitable for use in the constructor.
+     *
+     * @return the repository instance
+     */
+    protected R getRepository() {
```

Comment:
Most of the time it is easier to init in a `@PostConstruct` so you find any problems immediately